src/librustc_target/spec/uefi_msvc_base.rs

```diff
@@ -37,25 +30,29 @@ pub fn opts() -> TargetOptions {
         // exit (default for applications).
         "/subsystem:efi_application".to_string(),
     ];
-    let mut pre_link_args = LinkArgs::new();
-    pre_link_args.insert(LinkerFlavor::Msvc, pre_link_args_msvc.clone());
-    pre_link_args.insert(LinkerFlavor::Lld(LldFlavor::Link), pre_link_args_msvc);
+    base.pre_link_args.get_mut(&LinkerFlavor::Msvc).unwrap().extend(pre_link_args_msvc.clone());
+    base.pre_link_args
+        .get_mut(&LinkerFlavor::Lld(LldFlavor::Link))
+        .unwrap()
+        .extend(pre_link_args_msvc);
 
     TargetOptions {
-        dynamic_linking: false,
-        executables: true,
         disable_redzone: true,
         exe_suffix: ".efi".to_string(),
         allows_weak_linkage: false,
         panic_strategy: PanicStrategy::Abort,
         stack_probes: true,
         singlethread: true,
-        emit_debug_gdb_scripts: false,
-
         linker: Some("rust-lld".to_string()),
-        lld_flavor: LldFlavor::Link,
-        pre_link_args,
+        // FIXME: This should likely be `true` inherited from `msvc_base`
+        // because UEFI follows Windows ABI and uses PE/COFF.
+        // The `false` is probably causing ABI bugs right now.
+        is_like_windows: false,
+        // FIXME: This should likely be `true` inherited from `msvc_base`
+        // because UEFI follows Windows ABI and uses PE/COFF.
+        // The `false` is probably causing ABI bugs right now.
+        is_like_msvc: false,
```

Comment:
I am not even sure what these two are supposed to mean. They show up in the rustc code all over the place, but there is no real explanation to what they exactly control. When I wrote the UEFI target, I checked all their usages and was quite sure they do not effect any behavior on UEFI. However, checking again now, either their usage has increased or I missed some places back them.

I guess we should set them to true? I am quite certain that either way we will end up with code not meant for UEFI (either now or in the future when someone picks that option to toggle one more knob).

Regardless, I think it is independent of your PR. So I am definitely fine with adding these 2 comments. Thanks a lot for digging through this!

Reply:
>They show up in the rustc code all over the place, but there is no real explanation to what they exactly control.

`is_like_msvc` and `is_like_windows` are some ad-hoc flags controlling target-dependent behavior (introduced in https://github.com/rust-lang/rust/pull/16156, https://github.com/rust-lang/rust/pull/25350).
Sometimes it may mean "use Windows ABI", sometimes "the produced executable is PE/COFF", sometimes "search libraries in directories specific to visual studio", or something else.

I'm not sure *all* of these aspects apply to UEFI, so enabling them may both fix something and break something.

These flags likely need to be split into multiple more specific flags, but nobody bothered to do it yet because we had only windows-msvc and windows-gnu targets until recently.